File: AdjoinableMPIexamples/Tapenade_F77/ex01trivial/programAMPI.f
C Hand-turned into AMPI. Not differentiated.
c   One proc just sends, one proc just receives

      SUBROUTINE head(v1, v2)
      implicit none
      include 'ampi/ampif.h'
      DOUBLE PRECISION v1, v2
      INTEGER world_rank, err_code
      DOUBLE PRECISION value

      call AMPI_Comm_rank(MPI_COMM_WORLD, world_rank, err_code)
      IF (world_rank.EQ.0) THEN
         value = SQRT(v1)
         call AMPI_Send(value, 1, AMPI_ADOUBLE_PRECISION, 1, 0,
     +        AMPI_TO_RECV, MPI_COMM_WORLD, err_code)
      ELSE IF (world_rank.EQ.1) THEN
         call AMPI_Recv(value, 1, AMPI_ADOUBLE_PRECISION, 0, 0,
     +        AMPI_FROM_SEND, MPI_COMM_WORLD,
     +        MPI_STATUS_IGNORE, err_code)
         v2 = SIN(value)
      ENDIF
      END

      PROGRAM main
      implicit none
      include 'ampi/ampif.h'
      INTEGER world_rank, err_code
      DOUBLE PRECISION x,y

      call AMPI_Init_NT(err_code)
      call AMPI_Comm_rank(MPI_COMM_WORLD, world_rank, err_code)
      IF (world_rank.EQ.0) THEN
         x=3.5
         print *,"process",world_rank," sets val  [3.50000000==]",x
         call head(x,y)
      ELSE IF (world_rank.EQ.1) THEN
         call head(x,y)
         print *,"process",world_rank," gets val  [0.95532692==]",y
      ENDIF
      call AMPI_Finalize_NT(err_code)
      END

File: AdjoinableMPIexamples/Tapenade_C/ex01trivial/programAMPI.c
/* Program ported to AMPI. Porting should be made by the user
 * in general (e.g. for pairedWith), but could be automated sometimes.
 * Should behave identical to program.c
 * One proc just sends, one proc just receives */

#include <stdio.h>
#include <math.h>
#include "ampi/ampi.h"

void head(double* v1, double *v2) {
  int world_rank;
  AMPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  double value ;
  if (world_rank == 0) {
    value = sqrt(*v1) ;
    AMPI_Send(&value, 1, MPI_DOUBLE, 1, 0, AMPI_TO_RECV, MPI_COMM_WORLD);
  } else if (world_rank == 1) {
    AMPI_Recv(&value, 1, MPI_DOUBLE, 0, 0, AMPI_FROM_SEND, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
    *v2 = sin(value) ;
  }
}

int main(int argc, char** argv) {
  AMPI_Init_NT(0,0);
  int world_rank;
  AMPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  double x,y;
  if (world_rank == 0) {
    x=3.5;
    printf(__FILE__ ": process %i sets val  [3.500000==] %f \n", world_rank, x);
    head(&x,&y);
  } else if (world_rank == 1) {
    head(&x,&y);
    printf(__FILE__ ": process %i gets val  [0.955327==] %f \n", world_rank, y);
  } 
  AMPI_Finalize_NT();
  return 0;
}
